h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1758,17 +1758,22 @@ private void compactMoveChunks(Iterable<Chunk> move) {
             // to the end of the file
             writeStoreHeader();
             sync();
+
+            long reservedAreaSize = getFileLengthInUse();
             for (Iterator<Chunk> iter = move.iterator(); iter.hasNext(); ) {
                 Chunk chunk = iter.next();
                 if (fileStore.predictAllocation(chunk.len) >= leftmostBlock) {
                     break;
                 }
-                moveChunk(chunk, false);
+                moveChunk(chunk, 0);
                 iter.remove();
             }
-
+            // we need to ensure that chunks moved within the following loop
+            // do not overlap with space just released by chunks moved above,
+            // if some of them happened to sit right at the end of used space,
+            // hence the need to reserve this area
             for (Chunk chunk : move) {
-                moveChunk(chunk, true);
+                moveChunk(chunk, reservedAreaSize);
             }
 
```

Comment:
Here, everything is good. A part moved to the left of file, and another moved to the right of the file.